File: 653-two-sum-iv-input-is-a-bst/two-sum-iv-input-is-a-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool inorder(TreeNode* root,unordered_set<int> &uset,int k){
        if (root == nullptr) return false;

        if (uset.count(k - (root -> val))) return true;
        uset.insert(root -> val);

        bool left = inorder(root -> left,uset,k);
        bool right = inorder(root -> right,uset,k);

        return (left || right);
    }

    bool findTarget(TreeNode* root, int k) {
        
        unordered_set<int> uset;
        return inorder(root,uset,k);
    }
};

File: 653-two-sum-iv-input-is-a-bst/two-sum-iv-input-is-a-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    TreeNode global;
    public boolean helper(TreeNode root, int target){
        if(root == null) return false;

        if(root.val == target) return true;
        if(root.val < target) return helper(root.right, target);
        else return helper(root.left, target);
    }
    public boolean find(TreeNode root, int k) {
        if(root == null) return false;
        if (k != ((root.val)*2)  && (helper(global,  k - root.val))) return true;
        return find(root.left,  k) || find(root.right,  k );

    }
    public boolean findTarget(TreeNode root, int k) {
        global = root;
        return find(root, k);

    }
}